core/messagechains/src/sha256.rs

```diff
@@ -0,0 +1,55 @@
+#[derive(Debug, Clone, PartialOrd, Ord, PartialEq, Eq, Hash)]
+pub struct Sha256MessageDigest(pub [u8; 32]);
+
+impl crate::MessageDigest for Sha256MessageDigest {}
+
+impl Default for Sha256MessageDigest {
+    fn default() -> Self {
+        Sha256MessageDigest([0; 32])
+    }
+}
+
+pub struct Sha256MessageHasher<D: crate::DeviceId>(sha2::Sha256, std::marker::PhantomData<D>);
+
+impl<D: crate::DeviceId> Sha256MessageHasher<D> {
+    pub fn new() -> Self {
+        use sha2::Digest;
+
+        Sha256MessageHasher(sha2::Sha256::new(), std::marker::PhantomData)
+    }
+}
+
+impl<D: crate::DeviceId> crate::MessageHasher<D> for Sha256MessageHasher<D> {
+    type Output = Sha256MessageDigest;
+
+    fn hash_message<'a, BD: std::borrow::Borrow<D>>(
+        &'a mut self,
+        prev_digest: Option<&Self::Output>,
+        recipients: &mut impl Iterator<Item = BD>,
+        message: &[u8],
+    ) -> Self::Output
+    where
+        D: 'a,
+    {
+        use sha2::Digest;
+
+        if let Some(digest) = prev_digest {
+            self.0.update(&[b'p', b'r', b'e', b'v']);
+            self.0.update(&digest.0);
```

Comment:
We tag individual inputs of hash functions here, although this might well not be required. Generally, it's good practice to insert markers when calculating hashes over user-defined inputs, as users may be able to force a plaintext (and hence also hash) collision when crafting malicious input data. Given that all of the components (except for the last one) are of a fixed length here, it shouldn't matter much, except in one case: without the explicit `b'message'` marker, a message may be crafted that collides with another + a prefix of recipients. This doesn't really fit into our system model where the message & validation payload is part of an encrypted+authenticated payload and the server maintains the recipient list. However, we can just insert these markers and not worry about these hypothetical attack vectors any more.